datafusion/src/sql/planner.rs

```diff
@@ -213,9 +213,19 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
     ) -> Result<LogicalPlan> {
         let set_expr = &query.body;
         if let Some(with) = &query.with {
+            // a `WITH` block can't use the same name for many times
+            let mut cte_names = HashSet::new();
```

Comment:
Since this function gets a hashtable called `ctes` as a parameter, I wonder if you could check for the name in that hashtable rather than making a new one?

```rust
        ctes: &mut HashMap<String, LogicalPlan>,
```

Reply:
Oh, good catch, we can reuse `ctes`!